File: 2021/3/1.go
package main

import (
	"bytes"
	"fmt"
	"os"
)

func input() [][]byte {
	file, _ := os.ReadFile("./input")
	return bytes.Split(file, []byte("\n"))
}

func main() {
	input := input()

	lineLength := len(input[0])
	var bitFrequency []uint = make([]uint, lineLength)
	var gamma, epsilon, mask uint = 0, 0, (1<<lineLength - 1)

	for _, bits := range input {
		for i := 0; i < lineLength; i++ {
			bitFrequency[i] += uint(bits[i] - 48)
		}
	}

	for i := 0; i < len(bitFrequency); i++ {
		gamma = gamma << 1
		gamma += (bitFrequency[i] << 1) / uint(len(input))
	}

	epsilon = mask &^ gamma
	fmt.Printf("%b\n", gamma)
	fmt.Printf("%b\n", epsilon)
	fmt.Printf("%v\n", gamma*epsilon)
}

File: 2021/3/1.py
raw_data = open("power_cons.txt", "r").read()

ordered_data = raw_data.split()

lineLength = len(ordered_data[0])
bitFrequency = [0 for _ in ordered_data[0]]
gamma = 0
epsilon = 0
mask = (1 << lineLength) - 1

for line in ordered_data:
    for column in range(lineLength):
      bitFrequency[column] += int(line[column])

for column in bitFrequency:
  gamma = gamma << 1
  gamma += (column << 1) / len(ordered_data)

epsilon = mask ^ gamma

print(bitFrequency)
print("{0:b}".format(gamma))
print("{0:b}".format(epsilon))
print(gamma * epsilon)